File: apps/api-dotnet-archive/MedManager.Api/Auth/ExpiryNotifier.cs
using System.Text.Json;
using MedManager.Api.Data;
using Microsoft.EntityFrameworkCore;
using Microsoft.Extensions.Options;
using WebPush;

namespace MedManager.Api.Auth;

public class ExpiryNotifier(
    IServiceProvider services,
    IOptionsMonitor<PushOptions> pushOptions,
    ILogger<ExpiryNotifier> logger)
    : BackgroundService
{
    private static readonly int[] NotifyDaysBefore = [30, 7, 1, 0];
    private static readonly TimeOnly RunAt = new(9, 0);

    protected override async Task ExecuteAsync(CancellationToken stoppingToken)
    {
        while (!stoppingToken.IsCancellationRequested)
        {
            try
            {
                var delay = DelayUntilNextRun();
                await Task.Delay(delay, stoppingToken);
                if (!pushOptions.CurrentValue.IsConfigured)
                {
                    logger.LogDebug("Push not configured; skipping expiry notifier tick");
                    continue;
                }
                await RunAsync(stoppingToken);
            }
            catch (OperationCanceledException) { break; }
            catch (Exception ex)
            {
                logger.LogError(ex, "Expiry notifier tick failed");
            }
        }
    }

    private static TimeSpan DelayUntilNextRun()
    {
        var now = DateTime.UtcNow;
        var todayRun = now.Date.Add(RunAt.ToTimeSpan());
        var next = now < todayRun ? todayRun : todayRun.AddDays(1);
        return next - now;
    }

    private async Task RunAsync(CancellationToken ct)
    {
        using var scope = services.CreateScope();
        var db = scope.ServiceProvider.GetRequiredService<AppDbContext>();

        var today = DateOnly.FromDateTime(DateTime.UtcNow);
        var upperBound = today.AddDays(NotifyDaysBefore.Max());

        var candidates = await db.Medicines
            .Where(m => m.Quantity > 0 && m.ExpiryDate <= upperBound)
            .Include(m => m.User)
            .ToListAsync(ct);

        foreach (var medicine in candidates)
        {
            var effective = medicine.EffectiveExpiry;
            var daysLeft = effective.DayNumber - today.DayNumber;
            if (!NotifyDaysBefore.Contains(daysLeft)) continue;

            var subs = await db.PushSubscriptions
                .Where(s => s.UserId == medicine.UserId)
                .ToListAsync(ct);
            if (subs.Count == 0) continue;

            var locale = medicine.User.Locale;
            var payload = JsonSerializer.Serialize(new
            {
                title = locale == "en" ? "Medicine expiry reminder" : "İlaç son kullanma uyarısı",
                body = locale == "en"
                    ? $"{medicine.Name} — expires in {daysLeft} days."
                    : $"{medicine.Name} — {daysLeft} gün içinde sona eriyor.",
                url = $"/medicines/{medicine.Id}",
            });

            foreach (var sub in subs)
            {
                try
                {
                    await SendAsync(sub, payload);
                }
                catch (WebPushException ex) when (ex.StatusCode is System.Net.HttpStatusCode.Gone
                                                  or System.Net.HttpStatusCode.NotFound)
                {
                    db.PushSubscriptions.Remove(sub);
                }
                catch (Exception ex)
                {
                    logger.LogWarning(ex, "Push send failed for sub {SubId}", sub.Id);
                }
            }
        }

        await db.SaveChangesAsync(ct);
    }

    private Task SendAsync(Domain.PushSubscriptionEntity sub, string payload)
    {
        var opts = pushOptions.CurrentValue;
        var vapid = new VapidDetails(opts.Subject, opts.PublicKey!, opts.PrivateKey!);
        var webPush = new WebPushClient();
        var target = new WebPush.PushSubscription(sub.Endpoint, sub.P256dh, sub.Auth);
        return webPush.SendNotificationAsync(target, payload, vapid);
    }
}

File: apps/api/src/workers/expiryNotifier.ts
import { eq, lte } from "drizzle-orm";
import webpush from "web-push";
import { config } from "../config.js";
import { db } from "../db/client.js";
import { medicines, pushSubscriptions, users } from "../db/schema.js";
import { effectiveExpiryOf } from "../lib/medicines.js";

const NOTIFY_DAYS = [30, 7, 1, 0];

function configureWebPush() {
  if (!config.push.configured) return false;
  webpush.setVapidDetails(config.push.subject, config.push.publicKey, config.push.privateKey);
  return true;
}

export async function runExpiryTick(now = new Date()): Promise<{ sent: number; pruned: number }> {
  if (!configureWebPush()) return { sent: 0, pruned: 0 };

  const todayIso = now.toISOString().slice(0, 10);
  const upperBound = new Date(now.getTime() + Math.max(...NOTIFY_DAYS) * 24 * 60 * 60 * 1000)
    .toISOString()
    .slice(0, 10);

  const candidates = await db
    .select({ medicine: medicines, user: users })
    .from(medicines)
    .innerJoin(users, eq(users.id, medicines.userId))
    .where(lte(medicines.expiryDate, upperBound));

  let sent = 0;
  let pruned = 0;

  for (const { medicine, user } of candidates) {
    if (Number(medicine.quantity) <= 0) continue;
    const effective = effectiveExpiryOf(
      medicine.expiryDate,
      medicine.openedAt,
      medicine.openedShelfLifeDays,
    );
    const daysLeft = Math.round(
      (new Date(effective).getTime() - new Date(todayIso).getTime()) / (24 * 60 * 60 * 1000),
    );
    if (!NOTIFY_DAYS.includes(daysLeft)) continue;

    const subs = await db
      .select()
      .from(pushSubscriptions)
      .where(eq(pushSubscriptions.userId, medicine.userId));
    if (subs.length === 0) continue;

    const isEn = user.locale === "en";
    const payload = JSON.stringify({
      title: isEn ? "Medicine expiry reminder" : "İlaç son kullanma uyarısı",
      body: isEn
        ? `${medicine.name} — expires in ${daysLeft} days.`
        : `${medicine.name} — ${daysLeft} gün içinde sona eriyor.`,
      url: `/medicines/${medicine.id}`,
    });

    for (const sub of subs) {
      try {
        await webpush.sendNotification(
          { endpoint: sub.endpoint, keys: { p256dh: sub.p256dh, auth: sub.auth } },
          payload,
        );
        sent += 1;
      } catch (err: unknown) {
        const statusCode = (err as { statusCode?: number } | undefined)?.statusCode;
        if (statusCode === 404 || statusCode === 410) {
          await db.delete(pushSubscriptions).where(eq(pushSubscriptions.id, sub.id));
          pruned += 1;
        }
      }
    }
  }

  return { sent, pruned };
}

export function startExpiryNotifier() {
  if (!config.push.configured) {
    console.log("[expiryNotifier] Skipped: PUSH_PUBLIC_KEY/PRIVATE_KEY not set.");
    return () => undefined;
  }

  const runAtHour = 9;
  let timer: NodeJS.Timeout | null = null;

  const schedule = () => {
    const now = new Date();
    const target = new Date(now);
    target.setUTCHours(runAtHour, 0, 0, 0);
    if (target <= now) target.setUTCDate(target.getUTCDate() + 1);
    const delay = target.getTime() - now.getTime();
    timer = setTimeout(async () => {
      try {
        const result = await runExpiryTick();
        console.log(`[expiryNotifier] sent=${result.sent} pruned=${result.pruned}`);
      } catch (err) {
        console.error("[expiryNotifier] failed", err);
      }
      schedule();
    }, delay);
  };

  schedule();
  return () => {
    if (timer) clearTimeout(timer);
  };
}
